<a href="https://colab.research.google.com/github/BasilaThasni/Supervised_Learning_BreastCancerPrediction_ML/blob/main/BreastCancer_SupervisedLearning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#1.LOADING AND PREPROCESSING:

##Load the breast cancer dataset from sklearn.

In [ ]:
from sklearn.datasets import load_breast_cancer
cancer_data = load_breast_cancer()

In [ ]:
type(cancer_data)

sklearn.utils._bunch.Bunch

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt
from sklearn.datasets import load_breast_cancer
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.metrics import classification_report, f1_score, accuracy_score


In [ ]:
print(cancer_data)

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 

In [ ]:
cancer_data.data.shape

(569, 30)

In [ ]:
cancer_data.feature_names

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

In [ ]:
cancer_data.target_names

array(['malignant', 'benign'], dtype='<U9')

In [ ]:
cancer_data.data #show values in data

array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]])

In [ ]:
cancer_data.target #shows values in target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

In [ ]:
#Assign values of data and target to two different variables:
x = cancer_data.data
y = cancer_data.target

##Preprocess the data to handle any missing values and perform necessary feature scaling.

In [ ]:
np.isnan(x).sum()

0

In [ ]:
np.isnan(y).sum()

0

##NO MISSING VALUES IN THE DATA.


#Splitting the data for training and testing

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2,random_state=45)

##Feature scaling (StandardScaler)

In [ ]:
std_sclr = StandardScaler()
x_train = std_sclr.fit_transform(x_train)
x_test = std_sclr.transform(x_test)
print(f"x_train  after scaling = {x_train}\n \n x_test after scaling = {x_test}")

x_train  after scaling = [[-0.67979189  0.53343366 -0.71678383 ... -1.18488657 -0.75807409
  -1.00600259]
 [-0.74290307 -2.04523361 -0.75795272 ... -1.03367369  0.72615917
  -0.53963041]
 [-0.72569093  1.21839216 -0.73716035 ... -1.0185524   0.49972442
  -0.90855168]
 ...
 [-0.51340787  0.68037978 -0.50345411 ... -0.50760407 -1.01397638
   0.34278217]
 [-0.0744983   0.32012479 -0.09093349 ... -0.71159025 -1.04344392
   0.0268699 ]
 [ 0.28695663  2.47928462  0.19766461 ... -0.75241773  0.48266426
  -1.21910951]]
 
 x_test after scaling = [[-0.9064184  -0.50940972 -0.8411222  ...  0.6047179  -0.59212534
   0.58373221]
 [ 0.90659367  0.65904889  0.93329867 ...  0.56389042  0.22521106
  -0.51285818]
 [ 1.77293804  1.8275075   1.70261636 ...  0.81944019  0.35548858
  -0.23710425]
 ...
 [-0.47898359 -1.54514281 -0.54587055 ... -1.34139191 -0.9953653
  -0.75702088]
 [ 0.6111186   0.59979643  0.64636396 ... -0.16344355 -0.28814444
  -0.2617347 ]
 [ 1.76720066 -1.18725792  1.79410279 ...  0.710

#2.CLASSIFICATION ALGORITHM IMPLEMENTATION:

##Implement the following five classification algorithms:
##1. *Logistic Regression*
##2. *Decision Tree Classifier*
##3. *Random Forest Classifier*
##4. *Support Vector Machine (SVM)*
##5. *k-Nearest Neighbors (k-NN)*

## Using Logistic Regression:

In [ ]:
log_reg = LogisticRegression()
log_reg.fit(x_train, y_train)
y_pred_log_reg = log_reg.predict(x_test)
print(f"LogisticRegression:\n\n{classification_report(y_test, y_pred_log_reg)}\n\n")

LogisticRegression:

              precision    recall  f1-score   support

           0       1.00      0.95      0.98        42
           1       0.97      1.00      0.99        72

    accuracy                           0.98       114
   macro avg       0.99      0.98      0.98       114
weighted avg       0.98      0.98      0.98       114





## Using Decision Tree Classifier:

In [ ]:
tree_cls = DecisionTreeClassifier()
tree_cls.fit(x_train, y_train)
y_pred_tree_cls = tree_cls.predict(x_test)
print(f"DecisionTreeClassifier:\n\n{classification_report(y_test, y_pred_tree_cls)}\n\n")

DecisionTreeClassifier:

              precision    recall  f1-score   support

           0       0.92      0.83      0.88        42
           1       0.91      0.96      0.93        72

    accuracy                           0.91       114
   macro avg       0.91      0.90      0.90       114
weighted avg       0.91      0.91      0.91       114





##Using Random Forest Classifier:

In [ ]:
rand_frst = RandomForestClassifier()
rand_frst.fit(x_train, y_train)
y_pred_rand_frst = rand_frst.predict(x_test)
print(f"RandomForestClassifier:\n\n{classification_report(y_test, y_pred_rand_frst)}\n\n")

RandomForestClassifier:

              precision    recall  f1-score   support

           0       0.97      0.93      0.95        42
           1       0.96      0.99      0.97        72

    accuracy                           0.96       114
   macro avg       0.97      0.96      0.96       114
weighted avg       0.97      0.96      0.96       114





##Using Support Vector Machine:

In [ ]:
svm = SVC()
svm.fit(x_train, y_train)
y_pred_svm = svm.predict(x_test)
print(f"svm:\n\n{classification_report(y_test, y_pred_svm)}\n\n")

svm:

              precision    recall  f1-score   support

           0       1.00      0.93      0.96        42
           1       0.96      1.00      0.98        72

    accuracy                           0.97       114
   macro avg       0.98      0.96      0.97       114
weighted avg       0.97      0.97      0.97       114





##Using k-Nearest Neighbors:

In [ ]:
knn = KNeighborsClassifier()
knn.fit(x_train, y_train)
y_pred_knn = knn.predict(x_test)
print(f"knn:\n\n{classification_report(y_test, y_pred_knn)}\n\n")

knn:

              precision    recall  f1-score   support

           0       0.97      0.90      0.94        42
           1       0.95      0.99      0.97        72

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114





#COMPARISON BETWEEN MODELS:

# Model Performance:(weighted avg):
##**lOGISTIC REGRESSION:**

## precision: 0.98
## recall   : 0.98
## f1 score : 0.98



##**DECISIONTREECLASSIFIER:**

## precision: 0.91
## recall   : 0.91
## f1 score : 0.91



##**RANDOMFORESTCLASSIFIER:**

## precision: 0.97
## recall   : 0.96
## f1 score : 0.96



##**SUPPORTVECTORMACHINE:**

## precision: 0.97
## recall   : 0.97
## f1 score : 0.97




##**KNEIGHBORCLASSIFIER:**

## precision: 0.96
## recall   : 0.96
## f1 score : 0.96

#**From the above comparison,we see that the best model is by 'Logistic Regression' as the performance is higher compared to the others (0.98),98% of accuracy.**

#**The least accuracy is by Decision Tree Classifier with only 0.91,only 91% of accuracy.**

#This shows that the **logistic Regression** algorithm is the best to perform.
#The worst performing algorithm is **Decision Tree Classifier**.